source/extensions/filters/network/kafka/messages/offset_commit.h

```diff
@@ -131,156 +131,48 @@ class OffsetCommitRequest : public Request {
-  }
-
-protected:
-  Int32Deserializer partition_;
-  Int64Deserializer offset_;
-  Int64Deserializer timestamp_;
-  NullableStringDeserializer metadata_;
-};
+// api version 0
 
-/**
- * Deserializes array of OffsetCommitPartition-s v0
- */
+// Deserializes bytes into OffsetCommitPartition (api version 0): partition, offset, metadata
+class OffsetCommitPartitionV0Buffer
+    : public CompositeDeserializerWith3Delegates<OffsetCommitPartition, Int32Deserializer, Int64Deserializer, NullableStringDeserializer> {};
+// Deserializes array of OffsetCommitPartition-s v0
 class OffsetCommitPartitionV0ArrayBuffer
     : public ArrayDeserializer<OffsetCommitPartition, OffsetCommitPartitionV0Buffer> {};
-
-/**
- * Deserializes array of OffsetCommitPartition-s v1
- */
-class OffsetCommitPartitionV1ArrayBuffer
-    : public ArrayDeserializer<OffsetCommitPartition, OffsetCommitPartitionV1Buffer> {};
-
-/**
- * Deserializes bytes into OffsetCommitTopic v0 (which is composed of topic name + array of v0
- * partitions)
- */
-class OffsetCommitTopicV0Buffer : public Deserializer<OffsetCommitTopic> {
-public:
-  size_t feed(const char*& buffer, uint64_t& remaining) {
-    size_t consumed = 0;
-    consumed += topic_.feed(buffer, remaining);
-    consumed += partitions_.feed(buffer, remaining);
-    return consumed;
-  }
-  bool ready() const { return partitions_.ready(); }
-  OffsetCommitTopic get() const { return {topic_.get(), partitions_.get()}; }
-
-protected:
-  StringDeserializer topic_;
-  OffsetCommitPartitionV0ArrayBuffer partitions_;
-};
-
-/**
- * Deserializes bytes into OffsetCommitTopic v1 (which is composed of topic name + array of v1
- * partitions)
- */
-class OffsetCommitTopicV1Buffer : public Deserializer<OffsetCommitTopic> {
-public:
-  size_t feed(const char*& buffer, uint64_t& remaining) {
-    size_t consumed = 0;
-    consumed += topic_.feed(buffer, remaining);
-    consumed += partitions_.feed(buffer, remaining);
-    return consumed;
-  }
-  bool ready() const { return partitions_.ready(); }
-  OffsetCommitTopic get() const { return {topic_.get(), partitions_.get()}; }
-
-protected:
-  StringDeserializer topic_;
-  OffsetCommitPartitionV1ArrayBuffer partitions_;
-};
-
-/**
- * Deserializes array of OffsetCommitTopic-s v0
- */
+// Deserializes bytes into OffsetCommitTopic (api version 0): topic name, partitions (v0)
+class OffsetCommitTopicV0Buffer
+    : public CompositeDeserializerWith2Delegates<OffsetCommitTopic, StringDeserializer, OffsetCommitPartitionV0ArrayBuffer> {};
+// Deserializes array of OffsetCommitTopic-s v0
 class OffsetCommitTopicV0ArrayBuffer
     : public ArrayDeserializer<OffsetCommitTopic, OffsetCommitTopicV0Buffer> {};
+// Deserializes bytes into OffsetCommitRequest (api version 0): group_id, topics (v0)
+class OffsetCommitRequestV0Deserializer
+    : public CompositeDeserializerWith2Delegates<OffsetCommitRequest, StringDeserializer, OffsetCommitTopicV0ArrayBuffer> {};
 
```

Comment:
one can write:
```
class OffsetCommitRequestV0Deserializer:
public CompositeDeserializerWith2Delegates<
  OffsetCommitRequest,
  StringDeserializer,
  ArrayDeserializer<
    OffsetCommitTopic,
    CompositeDeserializerWith2Delegates<
      OffsetCommitTopic,
      StringDeserializer,
      ArrayDeserializer<
        OffsetCommitPartition,
        CompositeDeserializerWith3Delegates<
          OffsetCommitPartition,
          Int32Deserializer,
          Int64Deserializer,
          NullableStringDeserializer
>>>>> {}
```
but I think it makes it even worse to understand

Reply:
Yes that is not good. At this point, please just make the code is readable as you think it can be, while trying to remove egregious templates. When this is done I will take another pass and offer different suggestions if I see them. It will be much easier to do this with all the other changes I have asked for such as comments, etc.

Reply:
The remaining big-template classes will be present only in `messages/*.h` files, as we need to define the structure somehow.
But in the end they are not very 

The end code is actually pretty similar to how Java client code looks like ( https://github.com/apache/kafka/blob/2.1/clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java#L58 ) but Java without templates forces them to use raw Objects flying around and (nicely hidden) downcasts just like here https://github.com/apache/kafka/blob/2.1/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L211 

Other changes like comments, splits, renames etc. have been already applied.

Reply:
@adamkotwasinski OK so the PR is ready for another pass?

Reply:
Yes

Reply:
Can you fix DCO before I start reviewing again?

Reply:
done